cpp/src/quantiles/quantiles_util.hpp

```diff
@@ -0,0 +1,154 @@
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+template <typename Result, typename T>
+CUDA_HOST_DEVICE_CALLABLE
+Result get_array_value(T const* devarr, size_type location)
+{
+    T result;
+#if defined(__CUDA_ARCH__)
+    result = devarr[location];
+#else
+    CUDA_TRY( cudaMemcpy(&result, devarr + location, sizeof(T), cudaMemcpyDeviceToHost) );
+#endif
+    return static_cast<Result>(result);
+}
+
+namespace interpolate {
+
+template <typename Result, typename T>
+CUDA_HOST_DEVICE_CALLABLE
+Result linear(T lhs, T rhs, double frac)
+{
+    // TODO: safe operation to avoid overflow/underflow
+    // double can fully represent int8-32 value range.
+    // Since the fractoin part of double is 52 bits,
+    // double cannot fully represent int64.
+    // The underflow will be occurs at converting int64 to double
+    // detail: https://github.com/rapidsai/cudf/issues/1417
+
+    double dlhs = static_cast<double>(lhs);
+    double drhs = static_cast<double>(rhs);
+    double one_minus_frac = 1.0 - frac;
+    return static_cast<Result>(one_minus_frac * dlhs + frac * drhs);
+}
+
+template <typename Result, typename T>
+CUDA_HOST_DEVICE_CALLABLE
+Result midpoint(T lhs, T rhs)
+{
+    // TODO: try std::midpoint (C++20) if available
+    double dlhs = static_cast<double>(lhs);
+    double drhs = static_cast<double>(rhs);
+    return static_cast<Result>(dlhs / 2 + drhs / 2);
+}
+
+template <typename Result>
+CUDA_HOST_DEVICE_CALLABLE
+Result midpoint(int64_t lhs, int64_t rhs)
+{
+    // caring to avoid integer overflow and underflow between int64_t and Result( double )
+    int64_t half = lhs / 2 + rhs / 2;
+    int64_t rest = lhs % 2 + rhs % 2;
+    return static_cast<Result>(static_cast<Result>(half) +
+                               static_cast<Result>(rest) * 0.5);
+}
+
+template <>
+CUDA_HOST_DEVICE_CALLABLE
+int64_t midpoint(int64_t lhs, int64_t rhs)
+{
+    // caring to avoid integer overflow
+    int64_t half = lhs / 2 + rhs / 2;
+    int64_t rest = lhs % 2 + rhs % 2;
+    int64_t result = half;
+
+    // rounding toward zero
+    result += ( half >= 0 && rest != 0 ) ? rest/2 : 0;
+    result += ( half <  0 && rest != 0 ) ? 1 : 0;
+
+    return result;
+}
+
+} // namespace interpolate
+
+struct quantile_index
+{
+    size_type lower;
+    size_type higher;
+    size_type nearest;
+    double fraction;
+
+    quantile_index(size_type count, double quantile)
+    {
+        quantile = std::min(std::max(quantile, 0.0), 1.0);
+
+        double val = quantile * (count - 1);
+        lower = std::floor(val);
+        higher = static_cast<size_type>(std::ceil(val));
+        nearest = static_cast<size_type>(std::nearbyint(val));
+        fraction = val - lower;
+    }
+};
+
+template<typename Result, typename Source>
+Result
+select_quantile(Source source,
```

Comment:
Docs please. Something to explain the kind of function object being requested. Also, consider naming it something like `accessor` instead of `Source`